pkg/nodebridge/node_bridge.go

```diff
@@ -0,0 +1,313 @@
+)
+
+type NodeBridge struct {
+	Logger             *logger.Logger
+	Client             inx.INXClient
+	ProtocolParameters *inx.ProtocolParameters
+	TangleListener     *TangleListener
+	Events             *Events
+
+	isSyncedMutex      sync.RWMutex
+	latestMilestone    *inx.MilestoneInfo
+	confirmedMilestone *inx.MilestoneInfo
+
+	enableTreasuryUpdates bool
+	treasuryOutputMutex   sync.RWMutex
+	latestTreasuryOutput  *inx.TreasuryOutput
+}
+
+type Events struct {
+	MessageSolid              *events.Event
+	ConfirmedMilestoneChanged *events.Event
+}
+
+func INXMessageMetadataCaller(handler interface{}, params ...interface{}) {
+	handler.(func(metadata *inx.MessageMetadata))(params[0].(*inx.MessageMetadata))
+}
+
+func INXMilestoneCaller(handler interface{}, params ...interface{}) {
+	handler.(func(metadata *inx.Milestone))(params[0].(*inx.Milestone))
+}
+
+func NewNodeBridge(ctx context.Context, client inx.INXClient, enableTreasuryUpdates bool, logger *logger.Logger) (*NodeBridge, error) {
+	logger.Info("Connecting to node and reading protocol parameters...")
+
+	retryBackoff := func(_ uint) time.Duration {
+		return 1 * time.Second
+	}
+
+	protocolParams, err := client.ReadProtocolParameters(ctx, &inx.NoParams{}, grpc_retry.WithMax(5), grpc_retry.WithBackoff(retryBackoff))
+	if err != nil {
+		return nil, err
+	}
+
+	nodeStatus, err := client.ReadNodeStatus(ctx, &inx.NoParams{})
+	if err != nil {
+		return nil, err
+	}
+
+	return &NodeBridge{
+		Logger:             logger,
+		Client:             client,
+		ProtocolParameters: protocolParams,
+		TangleListener:     newTangleListener(),
+		Events: &Events{
+			MessageSolid:              events.NewEvent(INXMessageMetadataCaller),
+			ConfirmedMilestoneChanged: events.NewEvent(INXMilestoneCaller),
+		},
+		latestMilestone:       nodeStatus.GetLatestMilestone(),
+		confirmedMilestone:    nodeStatus.GetConfirmedMilestone(),
+		enableTreasuryUpdates: enableTreasuryUpdates,
+	}, nil
+}
+
+func (n *NodeBridge) DeserializationParameters() *iotago.DeSerializationParameters {
+	return &iotago.DeSerializationParameters{
+		RentStructure: &iotago.RentStructure{
+			VByteCost:    n.ProtocolParameters.RentStructure.GetVByteCost(),
+			VBFactorData: iotago.VByteCostFactor(n.ProtocolParameters.RentStructure.GetVByteFactorData()),
+			VBFactorKey:  iotago.VByteCostFactor(n.ProtocolParameters.RentStructure.GetVByteFactorKey()),
+		},
+	}
+}
+
+func (n *NodeBridge) MilestonePublicKeyCount() int {
+	return int(n.ProtocolParameters.GetMilestonePublicKeyCount())
+}
+
+func (n *NodeBridge) KeyManager() *keymanager.KeyManager {
+	keyManager := keymanager.New()
+	for _, keyRange := range n.ProtocolParameters.GetMilestoneKeyRanges() {
+		keyManager.AddKeyRange(keyRange.GetPublicKey(), milestone.Index(keyRange.GetStartIndex()), milestone.Index(keyRange.GetEndIndex()))
+	}
+	return keyManager
+}
+
+func (n *NodeBridge) Start(ctx context.Context) {
+	go n.listenToConfirmedMilestone(ctx)
```

Comment:
How do we detect that the underlying stream failed?
Show we do proper error handling here with channels etc?
Or is this no problem at all?

Reply:
the context gets cancelled

Reply:
Lets do that in a clean way with #7